---
title: "Downstream ATAC-seq QC"
author: "Victoria Flanary"
date: "2025-01-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Objective
Run ATAC-seq QC metrics (e.g. TSS enrichment score and fragment size distribution)
using ChrAccR.

# Set-up
```{r load packages}
library(ChrAccR)
library(tidyverse)
library(ggpubr)
library(rstudioapi)
library(here)
library(styler)
library(lintr)
```

```{r config settings}
theme_set(muRtools::theme_nogrid())
```

```{r load data}
dsa <- loadDsAcc(
  here("Sen_13cRA_ATAC", "data", "DsAtacDataset")
)
```

# Sample Annotation Table
```{r}
str(getSampleAnnot(dsa))
```

# Number of Each Region Type
```{r}
region_table <- data.frame(
  region = getRegionTypes(dsa),
  num_region = c(
    getNRegions(dsa, "promoters"),
    getNRegions(dsa, "distal_regions")
  )
)

write_csv(
  region_table,
  here(
    "Sen_13cRA_ATAC", "processing", "07_downstream_qc",
    "num_region_table.csv"
  )
)
```

# Number of Fragments per Sample
```{r}
num_fragments <- getFragmentNum(dsa, getSamples(dsa))

num_fragments_df <- data.frame(
  sample = dsa@sampleAnnot[["sample"]],
  num_fragments = num_fragments,
  row.names = NULL
)

write_csv(
  num_fragments_df,
  here(
    "Sen_13cRA_ATAC", "processing", "07_downstream_qc",
    "num_fragments_per_sample.csv"
  )
)
```

# Fragment Size Distribution
```{r}
sample_list <- getSamples(dsa)

plot_list <- list()
for (i in seq_along(sample_list)) {
  p <- plotInsertSizeDistribution(
    dsa, sample_list[i]
  ) + ggtitle(sample_list[i])

  plot_list[[sample_list[i]]] <- p

  print(paste0("Fragment distribution plot for ", sample_list[i], " done"))
}

pdf(
  here(
    "Sen_13cRA_ATAC", "processing", "07_downstream_qc",
    "fragment_size_distribution_plots.pdf"
  ),
  height = 6, width = 9
)

ggarrange(plotlist = plot_list, ncol = 4, nrow = 2)

dev.off()
```

# TSS Enrichment
```{r prepare granges object with tss coordinates}
tss_gr <- muRtools::getAnnotGrl.gencode("gencode.v27")[["gene"]]
# gencode.v22 not supported
tss_gr <- tss_gr[elementMetadata(tss_gr)[, "gene_type"] == "protein_coding"]
tss_gr <- promoters(tss_gr, upstream = 0, downstream = 1)
```

```{r filter non-canonical chromosomes}
chromosomes <- paste0("chr", c(1:22, "X", "Y"))
tss_gr_filt <- tss_gr[seqnames(tss_gr) %in% chromosomes]
seqlevels(tss_gr_filt) <- intersect(seqlevels(tss_gr), chromosomes)
```

```{r}
# Initialize list for plots and data frame for TSS enrichment scores
plot_list <- list()
tsse_scores <- data.frame(
  sample = character(), tsse_score = numeric(),
  tsse_score_smoothed = numeric(), stringsAsFactors = FALSE
)

# Loop through sample list
for (i in seq_along(sample_list)) {
  # Compute TSS enrichment
  tsse <- getTssEnrichment(dsa, sample_list[i], tss_gr_filt)

  # Extract TSS enrichment value
  enrichment_score <- tsse$tssEnrichment
  smoothed_score <- tsse$tssEnrichment.smoothed

  # Add to the plot list
  plot_list[[sample_list[i]]] <- tsse$plot + ggtitle(sample_list[i])

  # Add score to the data frame
  tsse_scores <- rbind(
    tsse_scores,
    data.frame(
      sample = sample_list[i],
      tsse_score = enrichment_score,
      tsse_score_smoothed = smoothed_score
    )
  )

  # Print progress
  print(paste0("TSSE Plot for ", sample_list[i], " done"))
}

# Save plots to a PDF
pdf(
  here(
    "Sen_13cRA_ATAC", "processing", "07_downstream_qc",
    "tss_enrichment_plots.pdf"
  ),
  height = 6, width = 9
)
ggarrange(plotlist = plot_list, ncol = 4, nrow = 2)
dev.off()

# Save TSS enrichment scores to a CSV file
write.csv(
  tsse_scores,
  file = here("Sen_13cRA_ATAC", "processing", "07_downstream_qc", "tss_enrichment_scores.csv"),
  row.names = FALSE
)

print("TSS enrichment scores saved.")
```

# Genome-Wide Tn5 Insertion Counts
```{r region-by-sample matrix with Tn5 insertion counts}
# promoters
tn5_inserts_promoters <- getCounts(dsa, "promoters")

# distal_regions
tn5_inserts_distal <- getCounts(dsa, "distal_regions")

# distribution of tn5_insertion counts by region
pdf(
  here(
    "Sen_13cRA_ATAC", "processing", "07_downstream_qc",
    "tn5_insertion_counts_by_region.pdf"
  )
)

par(mfrow = c(2, 2))
hist(tn5_inserts_promoters, main = "Tn5 Inserts in Promoter Regions")
hist(tn5_inserts_distal, main = "Tn5 Inserts in Distal Regions")

dev.off()
```

# End of script
```{r style code}
script <- rstudioapi::getSourceEditorContext()$path
style_file(script)
lint(script)
```

```{r clean environment}
rm(list = ls())
gc()
```
